velox/functions/lib/Re2Functions.cpp

```diff
@@ -1741,22 +1769,47 @@ PatternMetadata PatternMetadata::relaxedSuffix(
       PatternKind::kRelaxedSuffix,
       fixedLength,
       std::move(fixedPattern),
-      std::move(subPatterns)};
+      std::move(subPatterns),
+      {}};
 }
 
 PatternMetadata PatternMetadata::substring(const std::string& fixedPattern) {
-  return {PatternKind::kSubstring, fixedPattern.length(), fixedPattern, {}};
+  return {PatternKind::kSubstring, fixedPattern.length(), fixedPattern, {}, {}};
+}
+
+PatternMetadata PatternMetadata::substrings(
+    std::vector<std::string> substrings) {
+  return {PatternKind::kSubstrings, 0, "", {}, std::move(substrings)};
+}
+
+std::vector<std::string> PatternMetadata::parseSubstrings(
+    const std::string_view& pattern) {
+  // Not support substrings-search with '_' for best performance.
+  static const re2::RE2 fullPattern(R"((%+[^%_#\\]+)+%+)");
+  static const re2::RE2 subPattern(R"((?:%+)([^%_#\\]+))");
+  re2::StringPiece full(pattern);
+  re2::StringPiece cur;
+  std::vector<std::string> substrings;
```

Comment:
Wondering is there is a perf regression for the case where this pattern is not constant. Also, how do we handle custom escape characters? Is there some code that disables this optimization for that case?

Reply:
> Wondering is there is a perf regression for the case where this pattern is not constant. Also, how do we handle custom escape characters? Is there some code that disables this optimization for that case?

yeah,  so the real problem is that determinePatternKind will cause  perf regression, I think it could not be called for not constant, why we add it before,  and I think it should be removed for no-constant case

Reply:
> I think the parse code is not more complex than the left code in determinePatternKind

I wonder if that's indeed the case. I'm seeing that new code is using 2 regexes to match the pattern, applies second pattern multiple times, builds up a std::vector of strings using push_back w/o reserve + copies the strings. It might be worth benchmarking to determine whether this would cause significant regression and if so take care of that.

Reply:
I assume that for non-constant-pattern  there is perf regression, so only add it for constant case.